REGRESSION (275262@main): Unable to play games in PBS Kids Games app

jernoble opened the pull request:
#### 98f2399f9450976e2a8a780e73f6b1263b18ac15
<pre>
REGRESSION (275262@main): Unable to play games in PBS Kids Games app
<a href="https://bugs.webkit.org/show_bug.cgi?id=270284">https://bugs.webkit.org/show_bug.cgi?id=270284</a>
<a href="https://rdar.apple.com/123811318">rdar://123811318</a>

Reviewed by Chris Dumez.

The SharedAudioDestinationAdapter&apos;s configuration queue would become stuck, waiting on
a semaphore which would never be triggered. This appears to be a latent bug in the
RemoteAudioDestinationProxy/Manager, but one that has dire implications now that
SharedAudioDestination depends on its Adapter&apos;s render() method being called.

Rather than block on a Semaphore to ensure the Adapter is configured before calling
the completion handlers, simply pass the completion handler into the Adapter&apos;s
inner destination&apos;s start() or stop() methods. Only when the Adapter is already started
(or when it&apos;s stopped and does not need to be restarted) should the completion
handler get called explicitly.

Now that blocking is not needed, neither is the Adapter&apos;s configuration queue; all
the setup for configuration can happen on the main thread.

* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::protectedWorkBus):
(WebCore::SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestinationAdapter::addRenderer):
(WebCore::SharedAudioDestinationAdapter::removeRenderer):
(WebCore::SharedAudioDestinationAdapter::configureRenderThread):
(WebCore::SharedAudioDestinationAdapter::callAllConfigurationHandlers):
(WebCore::SharedAudioDestinationAdapter::render):
(WebCore::SharedAudioDestinationAdapter::protectedConfigurationQueue): Deleted.
(WebCore::m_configurationSemaphore): Deleted.

Canonical link: <a href="https://commits.webkit.org/275542@main">https://commits.webkit.org/275542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ffc55da6bddb13bfeea20af84dc52c8b72539b39

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21121 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44497 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44690 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38216 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44410 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24332 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18487 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34872 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42677 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36255 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15804 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37289 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46123 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38314 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37616 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41519 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16919 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13940 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40079 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18538 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9432 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18598 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18183 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -168,47 +149,57 @@ void SharedAudioDestinationAdapter::configureRenderThread(CompletionHandler<void
-
-        m_configurationSemaphore.wait();
+    if (shouldSkipRendering) {
+        callOnMainThread([completionHandler = WTFMove(completionHandler)] () mutable {
```
you're already on the main thread, does it need to be dispatched again?

jernoble replied:
IMO, yes, so that the completion handler runs on the next run loop and not the current one.

jyavenard commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -61,31 +61,31 @@ class SharedAudioDestinationAdapter : public ThreadSafeRefCountedAndCanMakeThrea
-    // Only accessed on m_configurationQueue:
-    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+    using RenderVector = Vector<RefPtr<SharedAudioDestination>>;
+    RenderVector m_renderers;
```
it is unclear at a glance that this is only ever used on the main thread. a `WTF_GUARDED_BY_CAPABILITY(mainThread)` would increase readability (and the associated assertIsCurrent(isMainThread())

jyavenard commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -61,31 +61,31 @@ class SharedAudioDestinationAdapter : public ThreadSafeRefCountedAndCanMakeThrea
 
     // Only accessed on the audio thread:
-    Vector<RefPtr<SharedAudioDestination>> m_configuredRenderers;
+    RenderVector m_configuredRenderers;
```
can we have related `WTF_GUARDED_BY_CAPABILITY` for it then?

jernoble replied:
I don't think there's a way to say `WTF_GUARDED_BY_CAPABILITY(!mainThread);`. At least, I don't see any use of it like that.

jyavenard replied:
Do you have access to the audio thread? then assertIsCurrent(audioThread()) 

anyway, if it's on the audio thread, we're now always accessing it while copy the m_newRenderers ; a lock in the audio thread is usually a worry.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -124,91 +122,85 @@ SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter()
-        m_renderers.removeFirst(renderer);
-        configureRenderThread(WTFMove(completionHandler));
-    });
+    m_renderers.removeFirst(&renderer);
```
nit: maybe we can `ASSERT(!m_renderers.contains(&renderer));` after that, to be safe.

cdumez commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -218,11 +210,14 @@ void SharedAudioDestinationAdapter::render(AudioBus* sourceBus, AudioBus* destin
-            m_configurationSemaphore.signal();
+            callOnMainThread([this, protectedThis = Ref { *this }, oldRenderers = WTFMove(oldRenderers)] () mutable {
+                callAllConfigurationHandlers(true);
+                oldRenderers.clear();
```
Surely this is not needed. Seems like a no-op given that it will go out of scope right after and get destroyed anyway.

jernoble replied:
I had thought that if I don't do _something_ here, I'll get an "unused captured variable" error. Let me check to ensure that's not necessary. I also worried about the possibility that an unused-lambda-captured-parameter behaviors like an unused-function-move-parameter and returns to the calling context to be destroyed there. I thought it would just be safer if the vector storage was explicitly destroyed in the main thread.

jernoble replied:
No build errors when I omit it. Seems odd but ::shrug::. I'll remove the explicit `.clear()`.

cdumez replied:
The vector destructor runs so there is a side effect and that's enough to make the compiler happy. You'd get a build error if you passed a raw pointer or a builtin type (like `unsigned`) and didn't use it in the lambda.

Note for example that we capture `protectedThis` in lambdas all the time and rarely use it inside the lambda.

cdumez reviewed (approved):
LGTM with nits.

jernoble committed:
REGRESSION (275262@main): Unable to play games in PBS Kids Games app
https://bugs.webkit.org/show_bug.cgi?id=270284
rdar://123811318

Reviewed by Chris Dumez.

The SharedAudioDestinationAdapter's configuration queue would become stuck, waiting on
a semaphore which would never be triggered. This appears to be a latent bug in the
RemoteAudioDestinationProxy/Manager, but one that has dire implications now that
SharedAudioDestination depends on its Adapter's render() method being called.

Rather than block on a Semaphore to ensure the Adapter is configured before calling
the completion handlers, simply pass the completion handler into the Adapter's
inner destination's start() or stop() methods. Only when the Adapter is already started
(or when it's stopped and does not need to be restarted) should the completion
handler get called explicitly.

Now that blocking is not needed, neither is the Adapter's configuration queue; all
the setup for configuration can happen on the main thread.

* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::protectedWorkBus):
(WebCore::SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestinationAdapter::addRenderer):
(WebCore::SharedAudioDestinationAdapter::removeRenderer):
(WebCore::SharedAudioDestinationAdapter::configureRenderThread):
(WebCore::SharedAudioDestinationAdapter::callAllConfigurationHandlers):
(WebCore::SharedAudioDestinationAdapter::render):
(WebCore::SharedAudioDestinationAdapter::protectedConfigurationQueue): Deleted.
(WebCore::m_configurationSemaphore): Deleted.

Canonical link: https://commits.webkit.org/275542@main
